### PR TITLE
feat(torch) : Native torch C++ model interface + NBEATS architecture implementation for time-series 

### PR DESCRIPTION
**Pull request by fantes**
https://github.com/philipperemy/n-beats/

### CODE REVIEW

**Review by sileht — APPROVED** [review]

**Comment by beniz** [comment]
@fantes Expecting to see more native models being part of the C++ code, what would be your thoughts on having a `torch/native/templates` directory to host them all ?

**Comment by fantes** [comment]
> @fantes Expecting to see more native models being part of the C++ codebase, what would be your thoughts on having a `torch/native/templates` directory to host them all ?
in my mind, the torch/native directory was the place to put them all, because torch/native itself will contains nothing else than the .h  (or we could put native.h in torch/ and native templates in torch/native)

**Comment by beniz** [comment]
> But if your prefer a template dir, I am ok

Let's say templates are part of a type of distribution (could be modulated depending on what templates are needed) whereas native interface is more generic and part of the framework I imagine.

File: src/backends/torch/torchlib.cc [hunk 1]
```diff
@@ -126,7 +128,9 @@ namespace dd
 
   c10::IValue TorchModule::forward(std::vector<c10::IValue> source)
   {
-    if (_native) // native modules take only one tensor as input for now
+    if (_graph) // native modules take only one tensor as input for now
+      return _graph->forward(to_tensor_safe(source[0]));
+    if (_native)
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:133]: Can we have both, or do we want an `else` here ?

File: src/backends/torch/torchlib.cc [hunk 2]
```diff
@@ -178,6 +182,8 @@ namespace dd
 
   std::vector<Tensor> TorchModule::parameters()
   {
+    if (_graph)
+      return _graph->parameters();
     if (_native)
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:187]: Is an `else` needed ?

File: src/backends/torch/torchlib.cc [hunk 3]
```diff
@@ -211,37 +219,48 @@ namespace dd
...
   {
-    if (_native)
-      _native->eval();
+    if (_graph)
+      _graph->eval();
     if (_traced)
       _traced->eval();
     if (_classif)
       _classif->eval();
+    if (_native)
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:242]: Pointing the possible need of `else` statements here, may be inappropriate though, unsure.

File: src/backends/torch/torchlib.cc [hunk 4]
```diff
@@ -346,9 +366,11 @@ namespace dd
       }
 
     // Create the model
-    if (this->_mlmodel._traced.empty() && this->_mlmodel._proto.empty())
-      throw MLLibInternalException(
-          "Use of libtorch backend needs either traced net or protofile");
+    if (this->_mlmodel._traced.empty() && this->_mlmodel._proto.empty()
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:369]: I suggest we wrap this long condition into a function that should be easier to maintain in the future. Also, we may want to put a nice comment on why every condition is there. We've seen this growing uncontrolled before in the Caffe lib API !

File: src/backends/torch/torchlib.cc [hunk 5]
```diff
@@ -534,15 +561,30 @@ namespace dd
...
+          {
+            CSVTSTorchInputFileConn *ctip
+                = reinterpret_cast<CSVTSTorchInputFileConn *>(&inputc);
+            _module._native = std::make_shared<NBeats>(*ctip, _nbeats_stacks);
+            // load params if any after instanciation
+            if (!this->_mlmodel._native.empty())
+              torch::load(_module._native, this->_mlmodel._native, _device);
+            _module._native->to(_device);
+          }
+        if (_module._graph)
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:575]: `else` applicable ?

File: src/backends/torch/torchlib.cc [hunk 6]
```diff
@@ -568,6 +610,10 @@ namespace dd
     // logging parameters
     int64_t log_batch_period = 20;
 
+    if (ad_mllib.has("nbeats_stacks"))
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:613]: Maybe we'd like to have the torch native C++ template models to be accomodated in a function, so that when we add more, we have placeholders ready. So that we can keep the generic train calls as a list of high level function calls.

File: src/backends/torch/torchlib.cc [hunk 7]
```diff
@@ -725,12 +773,32 @@ namespace dd
               }
             else if (_timeserie)
               {
-                if (_loss.empty() || _loss == "L1" || _loss == "l1")
-                  loss = torch::l1_loss(y_pred, y);
-                else if (_loss == "L2" || _loss == "l2" || _loss == "eucl")
-                  loss = torch::mse_loss(y_pred, y);
+                if (_template.find("nbeats") != std::string::npos)
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:776]: Can we wrap this up somewhere dedicated to the `nbeats` template ? Like `nbeat_native_template_loss_selection()` that would follow some generic signature ? This would allows simplifying the adding of more native templates.

File: src/backends/torch/torchlib.cc [hunk 8]
```diff
@@ -898,21 +968,23 @@ namespace dd
 
     bool lstm_continuation = false;
     TInputConnectorStrategy inputc(this->_inputc);
+    if (_template.find("nbeats") != std::string::npos)
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:971]: I'm thinking having a `nbeats_native_template_inputconn_config()` function somewhere. Possibly into the template actually, i.e. outside of `torchlib`.

**Comment by beniz** [comment]
Aside from my small comments, my main remark would be that I believe `torchlib` should be independent from torch native templates we may provide. This meaning that we may want any configuration code touching a given template to be part of the native template code itself. 
Some higher level function call could then take place into `torchlib`. Factory/virtual/policy-based should do it, I'm keen on discussing lower level details actually.

File: src/backends/torch/torchlib.cc (hunk 4 above)
Comment (sileht) [reply src/backends/torch/torchlib.cc:369]: Instead of documentation I suggest using explicit variable name like:
```
bool  unsupported_model_configuration = this->_mlmodel._traced.empty() && this->_mlmodel._proto.empty()
```

File: src/backends/torch/torchlib.cc (hunk 1 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:133]: I followed what was previously done with traced net that can followed by a weight net (fully connected) ; This allows for free to chain nets, if you really know what you are doing (typically few things will need to be rewritten around input connector -> first net) 
For know we do not have both, and at some time in the future, this should be superseded by the graph abstraction layer, ie a torchModule will be a node , that can be of one type native / traced / or (sub) graph.
Morality : if you think having a else makes things more clear, I am ok ;)

File: src/backends/torch/torchlib.cc (hunk 2 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:187]: no else because of return :)    this contradicts previsous comment, if we want to chain net for free, we need to cat parameters and not only send first one

File: src/backends/torch/torchlib.cc (hunk 3 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:242]: see first comment : this allows to chain nets for free, unused ATM

File: src/backends/torch/torchlib.cc (hunk 5 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:575]: same as before, one could want to chain a native and a graph, but with current code they take same input

File: src/backends/torch/torchlib.cc (hunk 6 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:613]: agreed, but I am not sure how to do it ATM, do you have an idea ?

File: src/backends/torch/torchlib.cc (hunk 7 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:776]: I am doing this directly in nbeats class, ie there will be here _module._native->loss() here, what do you think ?

File: src/backends/torch/torchlib.cc (hunk 8 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:971]: I am puttinh this as virtual into module._native/ native_net / nbeats , native_net class is there exactlty for doing so

**Comment by fantes** [comment]
> Aside from my small comments, my main remark would be that I believe `torchlib` should be independent from torch native templates we may provide. This meaning that we may want any configuration code touching a given template to be part of the native template code itself.
> Some higher level function call could then take place into `torchlib`. Factory/virtual/policy-based should do it, I'm keen on discussing lower level details actually.

native_net class was put there for doing so: to declare virtual stuff and hide everything, I prefer the inheritance way than policy based way, in case we have a lot of native stuff, and native API should be relatively stable

File: src/backends/torch/torchlib.cc (hunk 4 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:369]: I will propose a mix of both , let me know after my next push

File: src/backends/torch/torchlib.cc (hunk 6 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:613]: ok, I got an idea, see my upcoming push

File: src/backends/torch/torchlib.cc [hunk 9]
```diff
@@ -754,7 +807,9 @@ namespace dd
                   }
                 try
                   {
+                    auto p = _module.parameters();
```
Comment (BynaryCobweb) [inline src/backends/torch/torchlib.cc:810]: p is unused ?

Comment (fantes) [reply src/backends/torch/torchlib.cc:810]: remains from debug code, removed

File: src/backends/torch/torchlib.cc [hunk 10]
```diff
@@ -534,15 +561,30 @@ namespace dd
...
+            _module._graph->finalize(td.sizes());
             // reload params after finalize
             if (!this->_mlmodel._traced.empty())
-              torch::load(_module._native, this->_mlmodel._traced, _device);
+              torch::load(_module._graph, this->_mlmodel._traced, _device);
+            _module._graph->to(_device);
+          }
+        if (_module._native)
+          {
             _module._native->to(_device);
```
Comment (BynaryCobweb) [inline src/backends/torch/torchlib.cc:588]: `module._native->to(_device)` called a second time (first time l.572)

It could be great to centralize calls of the `to()` method in `TorchModule`, unless we need to have some models (graph, native, traced) on cpu and some other on gpu. What do you think?

File: src/backends/torch/torchlib.cc [hunk 11]
```diff
@@ -1044,6 +1110,9 @@ namespace dd
           }
         else if (_timeserie)
           {
+            if (_module._native != nullptr)
+              output = _module._native->cleanup_output(output);
+
```
Comment (BynaryCobweb) [inline src/backends/torch/torchlib.cc:1115]: Is the call to `cleanup_output` specific to timeseries or can we put it outside of the "if"?

Comment (fantes) [reply src/backends/torch/torchlib.cc:1115]: done

File: src/backends/torch/torchlib.cc (hunk 10 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:588]: ok, almost done :)

File: tests/ut-torchapi.cc [hunk 12]
```diff
@@ -44,6 +44,11 @@ static std::string bert_train_repo
 static std::string bert_train_data
     = "../examples/torch/bert_training_torch_140_transformers_251/data/";
 
+static std::string sinus = "../examples/all/sinus/";
+
+static std::string iterations_nbeats_cpu = "100";
```
Comment (beniz) [inline tests/ut-torchapi.cc:49]: FYI nbeats tests are not built on CPU because they are enclosed within the scope of a `#if !defined(CPU_ONLY)` that was originally put in place to remove BERT training tests from clogging CPU builds. We may want to add nbeats training to CPU unit tests, what do you think ?

Comment (fantes) [reply tests/ut-torchapi.cc:49]: ok, done

File: src/backends/torch/torchlib.cc [hunk 13]
```diff
@@ -211,37 +249,48 @@ namespace dd
...
-    if (_native)
-      _native->train();
+    if (_graph)
+      _graph->train();
     if (_traced)
       _traced->train();
     if (_classif)
       _classif->train();
+    if (_native)
+      _native->eval();
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:285]: Should be `train()` ?

Comment (fantes) [reply src/backends/torch/torchlib.cc:285]: it should. nice catch!

File: src/backends/torch/torchlib.cc [hunk 14]
```diff
@@ -534,17 +591,25 @@ namespace dd
     try
       {
         inputc.transform(ad);
+
+        _module._native = std::shared_ptr<NativeModule>(
+            NativeFactory::from_template(_template, _template_params, inputc));
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:596]: In Caffe backend, we try to instantiate the network template at service init. Do we want to do the same here ? As a side question, probably the template could be instantiated before the input are even transformed ?

File: src/backends/torch/torchlib.cc (hunk 13 above)
Comment (beniz) [reply src/backends/torch/torchlib.cc]: was it even training ? ;)

File: src/backends/torch/torchlib.cc (hunk 14 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:596]: unfortunately, instanciating nbeats needs the forecast and backcast length to be known, and they are "timesteps" input connector parameter, which can be given only at train/predict time

Comment (fantes) [reply src/backends/torch/torchlib.cc:596]: this is because for lstm, timestep is only used at timeserie length and lstm simply unroll over this data dimension, but for nbeats this has to be statically fixed

File: src/backends/torch/torchlib.cc (hunk 13 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc]: mmmm

Comment (fantes) [reply src/backends/torch/torchlib.cc]: train() is passed to all module and they do what they want with it, ie switch on/off dropout batchnorm and so on, most models do not do anything with it

File: src/backends/torch/torchlib.cc [hunk 15]
```diff
@@ -534,17 +591,25 @@ namespace dd
...
             // reload params after finalize
             if (!this->_mlmodel._traced.empty())
-              torch::load(_module._native, this->_mlmodel._traced, _device);
-            _module._native->to(_device);
+              torch::load(_module._graph, this->_mlmodel._traced, _device);
           }
+        _module.to(_device);
       }
     catch (...)
       {
```
Comment (BynaryCobweb) [inline src/backends/torch/torchlib.cc:615]: Do we still need the `try { ... } catch (...) { throw; }` ? It seems cumbersome and could be removed with no consequences

Comment (fantes) [reply src/backends/torch/torchlib.cc:615]: @beniz this is inspired from the try catch around inputc.transform in caffelib https://github.com/jolibrain/deepdetect/blob/5e982df6b71b55e26715f71f304d534daa59dc7d/src/backends/caffe/caffelib.cc#L1352 , do you have any comments ?

Comment (beniz) [reply src/backends/torch/torchlib.cc:615]: `transform` throws by design, if this is what you ask.

File: src/backends/torch/torchlib.cc [hunk 16]
```diff
@@ -536,17 +593,25 @@ namespace dd
...
         inputc.transform(ad);
+
+        _module._native = std::shared_ptr<NativeModule>(
+            NativeFactory::from_template(_template, _template_params, inputc));
+
         if (_module._native)
+          if (!this->_mlmodel._native.empty())
+            torch::load(_module._native, this->_mlmodel._native, _device);
+
+        if (_module._graph)
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:604]: I would recommend that anything that depends on values from within `_module` do actually happen in the `TorchModule` class. This is to avoid spreading the code and complexifying torchlib itself.

File: src/backends/torch/torchlib.cc [hunk 17]
```diff
@@ -900,21 +977,22 @@ namespace dd
...
+    if (_module._native != nullptr)
+      _module._native->update_input_connector(inputc);
+
     TOutputConnectorStrategy outputc(this->_outputc);
-    ;
     try
       {
         inputc.transform(ad);
-        if (_module._native)
+        if (_module._graph)
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:987]: Similar to my previous remark, I'd see this block of code going first into `TorchModule` with a function call replacing the block.

Next I'd envison that anything that deals with `graph` internals remains within `graph`, typically the `lstm_continues` stuff, i.e. as far as possible by design.

File: src/backends/torch/torchlib.cc [hunk 18]
```diff
@@ -212,37 +250,48 @@ namespace dd
...
       {
-        _native = std::make_shared<CaffeToTorch>(model._proto);
+        _graph = std::make_shared<CaffeToTorch>(model._proto);
         if (!model._traced.empty())
-          torch::load(_native, model._traced, _device);
+          torch::load(_graph, model._traced, _device);
+      }
+    if (!model._native.empty())
+      {
+        std::shared_ptr<NativeModule> m;
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:259]: My understanding is that at this stage, the native template hasn't been set. Before loading the model, we may want to put down a call to `NativeFactory::from_template` ?

File: src/backends/torch/torchlib.cc (hunk 14 above)
Comment (beniz) [reply src/backends/torch/torchlib.cc:598]: OK, maybe we can do it within `graph` ?

File: src/backends/torch/torchlib.cc (hunk 16 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:604]: ok, done

File: src/backends/torch/torchlib.cc (hunk 17 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:987]: ok, done

File: src/backends/torch/torchlib.cc (hunk 15 above)
Comment (BynaryCobweb) [reply src/backends/torch/torchlib.cc:617]: @beniz What do you mean by that? Is the `try ... catch` here as a remainder for exceptions thrown by `transform`?

File: src/backends/torch/torchlib.cc (hunk 18 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:259]: in case of native module/template, torchmodule::load should be called after factory::from_template, and it seems to me that it is the case in the code, i am not sure to understand your point

Comment (beniz) [reply src/backends/torch/torchlib.cc:316]: When `TorchModule::load` is called from `init_mllib`, it doesn't seem that at that stage `from_template` has been called. Actually since one call is mandatory before the other, we may want to regroup them into a function ?

Comment (fantes) [reply src/backends/torch/torchlib.cc:316]: in case of nbeats, it is not possible to instanciate model before having exact data size (length of timeseries), ie before the transform, so `init_mllib` cannot instanciate using `NativeFactory::from_template` . init_mllib loads a native net if available in repo, but it will be replaced by `NativeFactory::from_template` followed by another `TorchModule::load` at train or test time .
All we can do is remove this first model loading during init_mllib

File: src/backends/torch/torchlib.cc [hunk 19]
```diff
@@ -125,9 +127,98 @@ namespace dd
...
+    if (_graph)
+      _graph->to(device, dtype);
+    if (_native)
+      _native->to(device, dtype);
+    if (_traced)
+      _traced->to(device, dtype);
+  }
+
+  template <class TInputConnectorStrategy>
+  void TorchModule::post_transform(const std::string tmpl,
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:161]: we may call this function `load` for clarity ?

Comment (fantes) [reply src/backends/torch/torchlib.cc:161]: in my mind it was more like a hook (that loads in case of lstm and nbeats) or does nothing (in case of traced  becasue already loaded) . We should in the future clarify all this when unifying TorchModule as a graph

**Review by BynaryCobweb — APPROVED** [review]

File: src/backends/torch/torchlib.cc (hunk 19 above)
Comment (beniz) [reply src/backends/torch/torchlib.cc:161]: OK, great, to be clarified in terms of API in later updates!

**Review by beniz — APPROVED** [review]
